Return web build URL from successful `buildkite_trigger_build`

mokagio opened the pull request:
The idea came while working on automating the release for one of our apps. Having the URL of the web build the action scheduled available as the return value will enable fun things like annotating the trigger build with it.

## Checklist before requesting a review

- [x] Run `bundle exec rubocop` to test for code style violations and recommendations
- [x] Add Unit Tests (aka `specs/*_spec.rb`) if applicable
- [x] Run `bundle exec rspec` to run the whole test suite and ensure all your tests pass
- [x] Make sure you added an entry in [the `CHANGELOG.md` file](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/CHANGELOG.md#trunk) to describe your changes under the appropriate existing `###` subsection of the existing `## Trunk` section.
- [x] If applicable, add an entry in [the `MIGRATION.md` file](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/MIGRATION.md) to describe how the changes will affect the migration from the previous major version and what the clients will need to change and consider. – N.A.

mokagio committed:
Consistently use "build" over "job" in `buildkite_trigger_build_action`

"Build" is the correct word to use.
See https://buildkite.com/docs/pipelines/glossary#build

mokagio committed:
Return web build URL from successful `buildkite_trigger_build`

dangermattic commented:
<!--
  0 Errors
  3 Warnings: Please add an entry in the CHA...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please add an entry in the CHANGELOG.md file to describe the changes made by this PR</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">No reviewers have been set for this PR yet. Please request a review from <strong>@⁠wordpress-mobile/apps-infrastructure</strong>.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Add changelog entry for new `buildkite_trigger_build` return value

AliSoftware commented on spec/buildkite_trigger_build_action_spec.rb:
```diff
@@ -0,0 +1,48 @@
+    it 'returns the new build web URL' do
+      expected_url = 'https://fake.url'
+      allow(buildkit).to receive(:create_build)
+        .and_return(FakeResponse.new('scheduled', expected_url))
```
Wouldn't it be more idiomatic of `rspec` practices to return a `double(state: 'schedule', web_url: expected_url)` rather than creating our own `FakeResponse` type for the test?

_(Same remark on the rest of the test code of course)_

mokagio replied:
TIL about `double` and the [RuboCop-recommended](https://github.com/rubocop/rubocop-rspec/blob/863bafacd1a3313d707728fbba2623f8fcd9dd6b/lib/rubocop/cop/rspec/verified_doubles.rb#L6-L24) `instance_double`.

I left some notes about the resulting test implementation in 657c422. It's likely there's more Ruby/RSpec foo I'm not aware of that would address the minor issue I encountered there.

In the meantime, I'll merge the PR so the change is ready to ship in a new version. However, I won't be shipping a new version just yet. I have other unrelated things I'm working on and I might get to this later in the week. Just in case new PRs land in the meantime.

AliSoftware reviewed (approved):
🎉  this will come handy indeed! And nice to see more tests too 🙌 

Disabled auto-merge to let you address my small suggestion on the spec implementation, otherwise lgtm 👍

mokagio committed:
Use RSpec `double` instead of custom `FakeResponse`

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

Notice that this implementation is not ideal as far as RuboCop is
concerned.

The recommendation to use a verified double (`instance_double`) comes
with the intent of improving the test correctness by ensuring that the
expected object type is used in the system under test.

While that is an excellent general recommendation, in this particular
instance it backfires because the object `Buildkit` returns when
creating a build is `Sawyer::Resource` which we cannot as easily mock.

I.e., mocking with

```ruby
.and_return(
  instance_double(
    Sawyer::Resource,
    state: 'scheduled',
    web_url: expected_url
  )
)
```

Fails with:

```
the Sawyer::Resource class does not implement the instance method: state
```

I also tried to create a dummy `Sawyer::Resource` instance, but that
required providing a client object.

If this implementation is not satisfactory, we could consider stubbing
the Buildkite API HTTP response. I didn't go down that route when first
working on this because I didn't want to go look for endpoints and JSON
responses, but it's an option worth considering.

mokagio committed:
Add `buildkite_token` to `buildkite_trigger_build_action` in tests

The tests passed on my machine but failed in CI, see
https://buildkite.com/automattic/release-toolkit/builds/1602#018f0986-a43b-49d1-884d-50c62e8af094

It turns out I had a `BUILDKITE_TOKEN` in my environment which I had
forgotten about...
